### ISSUE TITLE
anti-xray not work

### ISSUE THREAD

**Issue by successed** [PaperMC/Paper#1749]
Paper.yml:

Anti-xray:
      enabled: true
      engine-mode: 1

İ use the paper-480 version.

bad.,,

check this img for anti-xray..

http://prntscr.com/lv2l6k

**Comment by successed** [comment]
1 - http://prntscr.com/lv2o4y
2- http://prntscr.com/lv2oc3

**Comment by successed** [comment]
http://prntscr.com/lv2pal

**Comment by successed** [comment]
paper.yml

http://prntscr.com/lv2phj

**Comment by zachbr** [comment]
Your issue is incomplete, at least tell us what plugins you’re using.

**Comment by successed** [comment]
plugin list 

http://prntscr.com/lv2q1q

**Comment by successed** [comment]
> Your issue is incomplete, at least tell us what plugins you’re using.

http://prntscr.com/lv2q1q

**Comment by mikroskeem** [comment]
FYI anti X-ray works in the picture, it's just on `engine-mode: 1`. Change it to 2 to make it obfuscate everything.

**Comment by zachbr** [comment]
Please try to make a test copy of your server either locally or elsewhere, and test without plugins, without protocol related plugins, etc. see if you can narrow it down to a certain set or interaction.

We are not going to be going through 60 plugins troubleshooting this issue

**Comment by successed** [comment]
> FYI anti X-ray works in the picture, it's just on `engine-mode: 1`. Change it to 2 to make it obfuscate everything.

this a soo laggy mean, i don't like this engine mode 2 t.y i tray mode 1...

**Comment by hugmanrique** [comment]
Each mode has its own compromises. If you feel you can improve either mode, feel free to PR your improvements.

**Comment by successed** [comment]
> Please try to make a test copy of your server either locally or elsewhere, and test without plugins, without protocol related plugins, etc. see if you can narrow it down to a certain set or interaction.
> 
> We are not going to be going through 60 plugins troubleshooting this issue

oke...

check img - http://prntscr.com/lv2v4h
check plugin list - http://prntscr.com/lv2vh6

**Comment by successed** [comment]
How can I make all of the blocks invisible? Diamond ore, emerald ore...

**Comment by successed** [comment]
http://prntscr.com/lv35p8

Chest is not invisible

**Comment by electronicboy** [comment]
If you're expecting all the bells and whistles, the built-in system is not suitable for your needs, the built-in anti x-ray system is designed as a lightweight alternative to more fully-fledged plugins

Chests are tile entities, not blocks, hence hiding them is more complex than just a standard block

Hiding blocks which are out of line of sight of the player but exposed to air will not be obfuscated, the alternative would be that we'd need to check every surface to see if a player can see them, which is not a simple job and is hardly amazing for performance

Would you please also write single messages instead of spamming the issue tracker with comments, we'd rather not have our email inboxes flooded

**Comment by k9lego** [comment]
I’ll write it out in plain English.
Here’s the issue that we’re facing. When we leave it on Engine 1 it hides about 75% of the items, but the ones exposed to caves, block gaps, etc are exposed to xrayers. This is easy to do because the check only needs to be made once then it can be cached. If you add line-of-sight that requires it to constantly be checking if the player can see the items. That’s causes lag. If you want that, get this plugin and disable Paper’s anti-X-ray (they’re incompatible): https://www.spigotmc.org/resources/orebfuscator.22818/ . Engine 2, however, is great for hiding ores. Hypixel uses this system for the UHC, I discovered it when I glitched out during a round. The issue with this is it causes FPS lag when all the blocks are updated. At a higher radius, there’s more lag. At a lower radius, you have a higher chance of seeing phantom ores. My proposal is simple, add a configurable delay for updating the blocks at 3-4+ radius so the clients don’t have to handle all the updates at once. At 1-2 blocks I’d leave it at the normal system to prevent phantom ores. Ideally there’d be configuration for all of this. Honestly this isn’t too important, but it would be helpful. I’ve gotten complaints of FPS dropping 25% or more on lower end PCs while mining.

**Comment by Ayymoss** [comment]
Same or similar issue.

See image for screenshot and config: https://i.imgur.com/EilJDDk.png

Any idea? The client I used was Aristois 1.13.2.

**Comment by stonar96** [comment]
@MrAmos123 The ores which you can see in your screenshots are ores which are either at chunk edges or above 64 blocks or exposed in caves. You can also hide those ores except for the third case.

The config option `chunk-edge-mode` controls the handling of the chunk edges. Please try `chunk-edge-mode: 2` or `3`. With `max-chunk-section-index` you can also configure the maximum height to which ores should be hidden. Please take a look at the PaperDocs https://paper.readthedocs.io/en/stable/server/configuration.html#anti-xray

@successed This might also solve your problem.

**Comment by Ayymoss** [comment]
I updated to latest build, just wondering if this is related... 

Note, this didn't happen with any older builds. I'm also unsure how I triggered it...

@zachbr @stonar96 (Apologies for pinging, I'm unsure how else you'd see this.)
```
[00:49:34 ERROR]: --- DO NOT REPORT THIS TO PAPER - THIS IS NOT A BUG OR A CRASH  - git-Paper-500 (MC: 1.13.2) ---
[00:49:34 ERROR]: The server has not responded for 10 seconds! Creating thread dump
[00:49:34 ERROR]: ------------------------------
[00:49:34 ERROR]: Server thread dump (Look for plugins here before reporting to Paper!):
[00:49:34 ERROR]: ------------------------------
[00:49:34 ERROR]: Current Thread: Server thread
[00:49:34 ERROR]: 	PID: 24 | Suspended: false | Native: false | State: TIMED_WAITING
[00:49:34 ERROR]: 	Stack:
[00:49:34 ERROR]: 		java.lang.Object.wait(Native Method)
[00:49:34 ERROR]: 		net.minecraft.server.v1_13_R2.PaperAsyncChunkProvider.requestChunk(PaperAsyncChunkProvider.java:211)
[00:49:34 ERROR]: 		net.minecraft.server.v1_13_R2.PaperAsyncChunkProvider.loadOrGenerateChunk(PaperAsyncChunkProvider.java:154)
[00:49:34 ERROR]: 		net.minecraft.server.v1_13_R2.PaperAsyncChunkProvider.getChunkAt(PaperAsyncChunkProvider.java:150)
[00:49:34 ERROR]: 		net.minecraft.server.v1_13_R2.ChunkProviderServer.getChunkAt(ChunkProviderServer.java:110)
[00:49:34 ERROR]: 		net.minecraft.server.v1_13_R2.PaperAsyncChunkProvider.getChunkAt(PaperAsyncChunkProvider.java:136)
[00:49:34 ERROR]: 		net.minecraft.server.v1_13_R2.World.getChunkAt(World.java:384)
[00:49:34 ERROR]: 		com.destroystokyo.paper.antixray.ChunkPacketBlockControllerAntiXray.onChunkPacketCreate(ChunkPacketBlockControllerAntiXray.java:164)
[00:49:34 ERROR]: 		net.minecraft.server.v1_13_R2.PlayerChunk.b(PlayerChunk.java:158)
[00:49:34 ERROR]: 		net.minecraft.server.v1_13_R2.PlayerChunkMap.c(PlayerChunkMap.java:226)
[00:49:34 ERROR]: 		net.minecraft.server.v1_13_R2.PlayerChunkMap.movePlayer(PlayerChunkMap.java:362)
[00:49:34 ERROR]: 		net.minecraft.server.v1_13_R2.PlayerList.updateChunks(PlayerList.java:417)
[00:49:34 ERROR]: 		net.minecraft.server.v1_13_R2.PlayerConnection.a(PlayerConnection.java:485)
[00:49:34 ERROR]: 		net.minecraft.server.v1_13_R2.PacketPlayInTeleportAccept.a(SourceFile:31)
[00:49:34 ERROR]: 		net.minecraft.server.v1_13_R2.PacketPlayInTeleportAccept.a(SourceFile:8)
[00:49:34 ERROR]: 		net.minecraft.server.v1_13_R2.PlayerConnectionUtils.lambda$ensureMainThread$0(PlayerConnectionUtils.java:13)
[00:49:34 ERROR]: 		net.minecraft.server.v1_13_R2.PlayerConnectionUtils$$Lambda$2217/1013604190.run(Unknown Source)
[00:49:34 ERROR]: 		java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[00:49:34 ERROR]: 		java.util.concurrent.FutureTask.run(FutureTask.java:266)
[00:49:34 ERROR]: 		net.minecraft.server.v1_13_R2.SystemUtils.a(SystemUtils.java:108)
[00:49:34 ERROR]: 		net.minecraft.server.v1_13_R2.MinecraftServer.b(MinecraftServer.java:1011)
[00:49:34 ERROR]: 		net.minecraft.server.v1_13_R2.DedicatedServer.b(DedicatedServer.java:439)
[00:49:34 ERROR]: 		net.minecraft.server.v1_13_R2.MinecraftServer.a(MinecraftServer.java:938)
[00:49:34 ERROR]: 		net.minecraft.server.v1_13_R2.MinecraftServer.run(MinecraftServer.java:836)
[00:49:34 ERROR]: 		java.lang.Thread.run(Thread.java:748)
[00:49:34 ERROR]: ------------------------------
[00:49:34 ERROR]: --- DO NOT REPORT THIS TO PAPER - THIS IS NOT A BUG OR A CRASH ---
[00:49:34 ERROR]: ------------------------------
```

**Comment by stonar96** [comment]
@MrAmos123 Yes looks like there is a problem. I have found out that this occurs first at Build 311 (https://github.com/PaperMC/Paper/commit/08bb5b49d25c7376c937bf790ef028d5de3fde41) with `chunk-edge-mode: 3`. Build 310 works fine. I am not very familiar with the async chunk api. @zachbr @aikar can you take a look at this?

**Comment by zachbr** [comment]
@MrAmos123 If you’re seeing that consistently please try using chunk-edge-mode: 2 instead. 

The async apis may still not be forgiving enough for this added loaded on servers, we will have to monitor the situation. I also know there were other tweaks being considered for the async system but I’m not aware of the state of them or if they’re applicable to this case.

@stonar96 I am not entirely familiar with that change specifically but will try and look into it more tomorrow afternoon.
